tests/infra/runner.py

```diff
@@ -200,3 +205,88 @@ def run(get_command, args):
                 for remote_client in clients:
                     remote_client.stop()
                 raise
+
+
+FAILURES = []
+
+
+def log_exception(args: threading.ExceptHookArgs):
+    description = f"Failure in {args.thread.name}: {repr(args.exc_value)}"
+    FAILURES.append(description)
+    LOG.error(
+        description
+        + "\n"
+        + "\n".join(
+            better_exceptions.format_exception(
+                args.exc_type, args.exc_value, args.exc_traceback
+            )
+        )
+    )
+
+
+threading.excepthook = log_exception
+
+
+class ConcurrentRunner:
+    threads: List[threading.Thread] = []
+
+    def __init__(self, add_options=None) -> None:
+        def add(parser):
+            parser.add_argument(
+                "-N",
+                help="List all sub-tests",
+                action="store_true",
+            )
+            parser.add_argument(
+                "-R", help="Run sub-tests whose name includes this string"
+            )
+            if add_options:
+                add_options(parser)
+
+        self.args = infra.e2e_args.cli_args(add=add)
+
+    def add(self, prefix, target, **args_overrides):
+        args_ = copy.deepcopy(self.args)
+        for k, v in args_overrides.items():
+            setattr(args_, k, v)
+        args_.label = f"{prefix}_{self.args.label}"
+        self.threads.append(threading.Thread(name=prefix, target=target, args=[args_]))
+
+    def run(self, max_concurrent=None):
+        config = {
+            "handlers": [
```

Comment:
I wonder whether `thread.name` should be at the very start of the line so that it's easier to read and parse?

Reply:
It seemed logical to group it with the module and function name, rather than put it on the other side of the timestamp, but I don't feel strongly either way, happy to move it left.

Reply:
I've just had a look at the CI output and it actually looks quite neat as it is!